Abort handler

Env 
Windows 10 Pro. Version 1809
RAM 32 gb
Proc Intel (R) Xeon(R) cpu E5-2686 v4 2.30GHz

Selenoid:  selenoid_windows_amd64.exe
Git Revision: 1.7.2
UTC Build Time: 2018-08-11_08:39:59AM

browsers 18 chrome 70 v, chromedriver 2.43

<img width="862" alt="screen shot 2018-11-02 at 11 23 06 am" src="https://user-images.githubusercontent.com/19333856/47914274-a68d8f80-dea7-11e8-8efd-73f08f650678.png">

Now reproducing on Linux too (at least in `1.8.1`).
```
2018/12/14 15:48:21 httputil: ReverseProxy read error during body copy: unexpected EOF
panic: net/http: abort Handler

goroutine 78808798 [running]:
net/http/httputil.(*ReverseProxy).ServeHTTP(0xc0055d8280, 0x821b60, 0xc0051ac620, 0xc00158f700)
        /home/travis/.gimme/versions/go1.11.1.linux.amd64/src/net/http/httputil/reverseproxy.go:284 +0x1058
main.proxy.func1(0xc00298ed20, 0x821b60, 0xc0051ac620, 0xc00158f700)
        /home/travis/gopath/src/github.com/aerokube/selenoid/selenoid.go:445 +0xfb
created by main.proxy
        /home/travis/gopath/src/github.com/aerokube/selenoid/selenoid.go:403 +0x7a
```

--- a/.travis.yml
+++ b/.travis.yml
@@ -2,7 +2,7 @@ sudo: required
 language: go
 
 go:
-  - 1.11.x
+  - 1.11.1
 
 services:
   - docker

--- a/selenoid.go
+++ b/selenoid.go
@@ -403,51 +403,51 @@ func generateRandomFileName(extension string) string {
 
 func proxy(w http.ResponseWriter, r *http.Request) {
 	done := make(chan func())
-	go func(w http.ResponseWriter, r *http.Request) {
-		cancel := func() {}
-		defer func() {
-			done <- cancel
-		}()
-		(&httputil.ReverseProxy{
-			Director: func(r *http.Request) {
-				requestId := serial()
-				fragments := strings.Split(r.URL.Path, slash)
-				id := fragments[2]
-				sess, ok := sessions.Get(id)
-				if ok {
-					sess.Lock.Lock()
-					defer sess.Lock.Unlock()
-					select {
-					case <-sess.TimeoutCh:
-					default:
-						close(sess.TimeoutCh)
-					}
-					if r.Method == http.MethodDelete && len(fragments) == 3 {
-						if enableFileUpload {
-							os.RemoveAll(filepath.Join(os.TempDir(), id))
-						}
-						cancel = sess.Cancel
-						sessions.Remove(id)
-						queue.Release()
-						log.Printf("[%d] [SESSION_DELETED] [%s]", requestId, id)
-					} else {
-						sess.TimeoutCh = onTimeout(sess.Timeout, func() {
-							request{r}.session(id).Delete(requestId)
-						})
-						if len(fragments) == 4 && fragments[len(fragments)-1] == "file" && enableFileUpload {
-							r.Header.Set("X-Selenoid-File", filepath.Join(os.TempDir(), id))
-							r.URL.Path = "/file"
-							return
-						}
-					}
-					r.URL.Host, r.URL.Path = sess.URL.Host, path.Clean(sess.URL.Path+r.URL.Path)
-					return
+	go func() {
+		(<-done)()
+	}()
+	cancel := func() {}
+	defer func() {
+		done <- cancel
+	}()
+	(&httputil.ReverseProxy{
+		Director: func(r *http.Request) {
+			requestId := serial()
+			fragments := strings.Split(r.URL.Path, slash)
+			id := fragments[2]
+			sess, ok := sessions.Get(id)
+			if ok {
+				sess.Lock.Lock()
+				defer sess.Lock.Unlock()
+				select {
+				case <-sess.TimeoutCh:
+				default:
+					close(sess.TimeoutCh)
 				}
-				r.URL.Path = errorPath
-			},
-		}).ServeHTTP(w, r)
-	}(w, r)
-	go (<-done)()
+				if r.Method == http.MethodDelete && len(fragments) == 3 {
+					if enableFileUpload {
+						os.RemoveAll(filepath.Join(os.TempDir(), id))
+					}
+					cancel = sess.Cancel
+					sessions.Remove(id)
+					queue.Release()
+					log.Printf("[%d] [SESSION_DELETED] [%s]", requestId, id)
+				} else {
+					sess.TimeoutCh = onTimeout(sess.Timeout, func() {
+						request{r}.session(id).Delete(requestId)
+					})
+					if len(fragments) == 4 && fragments[len(fragments)-1] == "file" && enableFileUpload {
+						r.Header.Set("X-Selenoid-File", filepath.Join(os.TempDir(), id))
+						r.URL.Path = "/file"
+						return
+					}
+				}
+				r.URL.Host, r.URL.Path = sess.URL.Host, path.Clean(sess.URL.Path+r.URL.Path)
+				return
+			}
+			r.URL.Path = errorPath
+		},
+	}).ServeHTTP(w, r)
 }
 
 func reverseProxy(hostFn func(sess *session.Session) string, status string) func(http.ResponseWriter, *http.Request) {

--- a/selenoid_test.go
+++ b/selenoid_test.go
@@ -354,6 +354,23 @@ func TestProxySession(t *testing.T) {
 	queue.Release()
 }
 
+func TestProxySessionPanicOnAbortHandler(t *testing.T) {
+
+	manager = &HTTPTest{Handler: Selenium()}
+
+	resp, err := http.Post(With(srv.URL).Path("/wd/hub/session"), "", bytes.NewReader([]byte("{}")))
+	AssertThat(t, err, Is{nil})
+	var sess map[string]string
+	AssertThat(t, resp, AllOf{Code{http.StatusOK}, IsJson{&sess}})
+
+	req, _ := http.NewRequest(http.MethodGet, With(srv.URL).Path(fmt.Sprintf("/wd/hub/session/%s/url?abort-handler=true", sess["sessionId"])), nil)
+	resp, err = http.DefaultClient.Do(req)
+	AssertThat(t, err, Not{nil})
+
+	sessions.Remove(sess["sessionId"])
+	queue.Release()
+}
+
 func TestSessionDeleted(t *testing.T) {
 	canceled := false
 	ch := make(chan bool)

--- a/utils_test.go
+++ b/utils_test.go
@@ -10,6 +10,7 @@ import (
 	"net/http"
 	"net/http/httptest"
 	"net/url"
+	"strconv"
 	"strings"
 	"sync"
 
@@ -119,6 +120,13 @@ func Selenium() http.Handler {
 		lock.RUnlock()
 		if !ok {
 			http.Error(w, "Session not found", http.StatusNotFound)
+			return
+		}
+		if r.FormValue("abort-handler") != "" {
+			out := "this call was relayed by the reverse proxy"
+			// Setting wrong Content-Length leads to abort handler error
+			w.Header().Add("Content-Length", strconv.Itoa(2*len(out)))
+			fmt.Fprintln(w, out)
 			return
 		}
 		d, _ := time.ParseDuration(r.FormValue("timeout"))

